### DISCUSSION TITLE
How can I avoid fetching data from my database twice, first to get the logged in auth.user, then to fetch their data from my own profile table where I keep their data?

### DISCUSSION THREAD

**Discussion by lumenwrites** [supabase/supabase#7488] [Questions]
Hi! I'm building a Next.js app with Supabase, using the help of `@supabase/auth-helpers-nextjs` library. I have two tables, `auth.users` managed by Supabase Auth, and `Profile`, where I keep all the custom data I want to save about the users.

Currently, to make the data from `Profile` available in my app, I have to make two requests to the database:
1. First one made by `useUser()` function from `'@supabase/auth-helpers-react'`, to fetch the id of the logged in user.
2. Second one is made by me, to use the user id to fetch the data from my profile info.

The first request seems unnecessary, since all I need is to get user.id from cookies managed by Supabase Auth. Is there any way for me to avoid making this request? Can you share some advice on how to correctly handle this?

Here's how my current data fetching code looks like, inside my auth context:

```javascript
  // Next helpers automatically fetch the logged in user info
  const { user: helpersUser, error } = useUser()
  // My own user, containing the info I fetched from the Profile table
  const [user, setUser] = useState()
  const [isLoading, setIsLoading] = useState(true)

  useEffect(() => {
    async function getUserProfile() {
      const { data: profile } = await supabaseClient
        .from('Profile')
        .select('*')
        .eq('id', helpersUser.id)
        .single()
      console.log('[AuthContext] Fetched profile data', profile)
      // Combine data from supabase auth table, with data from profile
      setUser({ ...helpersUser, ...profile })
      setIsLoading(false)
    }
    // Only fetch profile info once user is logged in.
    if (helpersUser) {
      getUserProfile()
    }
    if (error) {
      console.log('[AuthContext] Login error', error)
      setIsLoading(false)
    }
  }, [helpersUser, error])
```

I have to do something similar on the backend:
```javascript
export async function getServerSideProps({ params, req, res }) {
  const { user } = await getUser({ req, res }) // Access the user object

  if (user) { // if the user is logged in, fetch their profile info
    const profile = await prisma.profile.findUnique({
      where: { id: user.id },
      include: { boughtCourses: true },
    })
   [...]
  }
  [...]
}
```

Can you help me find the right way to avoid doing that as well?

**Comment by mansueli** [comment]
Have you tried to use Row Level Security to show the courses for the logged user?

Here are some resources:
[Our docs](https://supabase.com/docs/guides/auth/row-level-security)
[Cheat Sheet](https://supabase-cheatsheet.vercel.app/guide/advanced/row-level-security.html#security-invoker)

**Reply by lumenwrites** [reply]
Unfortunately this solution won't work for me, I need to be able to store and fetch custom data about my users, figuring out how to fetch the `Profile` table is important.

I mean, Supabase already stores some identifying user information in the browser somehow, right? In the cookies or something? I just need to find a way to get the user id out of that, and use that id to fetch `Profile` table in a single request.

@burggraf advised me to use [`auth.user()`](https://supabase.com/docs/reference/javascript/auth-user) function, but unfortunately it didn't work, this function takes the user info from memory, and it doesn't work when the user reloads the page.

On the server side, `@supabase/auth-helpers-nextjs` library has a `getUser(ctx)` function which somehow gets the user info from the cookies (I'm guessing) without reaching out to the database (I think).

There has to be some way to do this on the frontend, I'm just not familiar enough with supabase to find it, and google isn't helping.

**Reply by hf** [reply]
Hey @lumenwrites, what @mansueli is suggesting there will remove the need for getting the user information first. The reason is that if you use Row Level Security on your `Profile` table, it will be just like adding a `WHERE user_id = <logged in user>` clause to any `SELECT` statements there. So basically, your users will never be able to read anyone else's profile. 

I recommend you check out the links provided by @mansueli as they explain how this works very well and through a practical example.

**Reply by lumenwrites** [reply]
@hf Thank you very much for your advice! Can you clarify something please? Once I have implemented the RLS on my `Profile` table, which says that the user can only SELECT their own profile (`auth.uid()= id`), does that mean that I just have to fetch all the information from the `Profile` table, and it'll return just my own profile?

I'm trying to do this on my client:
```javascript
const { data: profile } = await supabaseClient.from('Profile').select('id, email, stripeCustomerId')
```
And it returns nothing, no data.

But if I use the `useUser()` hook from `@supabase/auth-helpers-react`, and make sure that it has done it's thing and fetched the user from `auth.users`, and then fetch the profile data, it works. I'm really confused as to why.

So I have implemented the RLS, but I still can't fetch the `Profile` data without telling the helpers function to go and fetch the user first.

**Comment by nattstack** [comment]
> The first request seems unnecessary, since all I need is to get user.id from cookies managed by Supabase Auth. Is there any way for me to avoid making this request? Can you share some advice on how to correctly handle this?

I believe to get the signed-in user id without the first request, you can do this instead:

```js
// https://supabase.com/docs/reference/javascript/auth-user
const user = supabase.auth.user() // for you it's supabaseClient.auth.user() I think
const userId = user.id;

const { data: profile } = await supabaseClient
...
```

**Reply by lumenwrites** [reply]
Thank you for your reply! Unfortunately, this doesn't work. `supabaseClient.auth.user()` returns the user when I sign in for the first time (as soon as I'm redirected to my website by Supabase Auth), but if I refresh the page after that, this function returns `null`.

Can you please help me to understand why?

**Reply by nattstack** [reply]
My guess is your are using `Next.js` and SSR is turn on by default. After refreshing, the user could be null because Supabase client reads from local storage for session and user information (I think) and local storage is being cleared on refresh.

**Reply by lumenwrites** [reply]
Yeah, SSR is turned on by default.

Do you know if there's any way to read the information from cookies? Or some other workaround to this?

Because if I can't get information after refreshing the page, this solution can't work for me, I need to get the user's info every time they open a website.

**Reply by nattstack** [reply]
Have you checked out this guide? 
https://supabase.com/docs/guides/with-nextjs

If you have and auth still doesn't work for you, I don't mind creating an example repo for you

**Reply by lumenwrites** [reply]
Yeah, I did follow this guide, and my auth does work. You can see my auth code [here](https://github.com/lumenwrites/gdschool/blob/main/context/AuthContext.tsx).

My problem is that every time user loads website, I have to make a request to the database twice.

First, on line 14, when I'm using the `useUser()` hook to fetch the signed in user:
```javascript
 const { user: helpersUser, error, isLoading: helpersLoading } = useUser()
```

And then on line 20, to fetch the `Profile` data associated with this user:
```javascript
      const { data: profile } = await supabaseClient
        .from('Profile')
        .select('id, email, stripeCustomerId')
        .eq('id', helpersUser.id)
        .single()
```

There's no reason for the first request to the database, I'm only doing it to get access to the `id` of the logged in user, so that I can use it to fetch the correct profile. What I'd really love some help with is figuring out how to fetch `Profile` info without fetching the logged in user first.

**Reply by JUNIORCO** [reply]
@lumenwrites did you figure this out?

**Reply by Metaphysics0** [reply]
I'm experiencing the same situation, let me know if you think of anything.

Maybe there's a way to specify only fetching the `id` from the initial auth call to make things slightly faster, but not sure how much that would help
